python/pyspark/sql/observation.py

```diff
@@ -0,0 +1,100 @@
+      the Dataset during the action.
+    - It will report the value of the defined aggregate columns as soon as we reach the end of
+      the action.
+
+    The metrics columns must either contain a literal (e.g. lit(42)), or should contain one or
+    more aggregate functions (e.g. sum(a) or sum(a + b) + avg(c) - lit(1)). Expressions that
+    contain references to the input Dataset's columns must always be wrapped in an aggregate
+    function.
+
+    Example:
+        >>> observation = Observation("my_metrics")
+        >>> observed_df = df.observe(observation, count(lit(1)).as("rows"), max($"id").as("maxid"))
+        >>> observed_df.write.parquet("ds.parquet")
+        >>> metrics = observation.get
+
+    An Observation instance collects the metrics while the first action is executed. Subsequent
+    actions do not modify the metrics returned by `Observation.get`. Retrieval of the metric via
+    `Observation.get` blocks until the first action has finished and metrics become available.
+
+    This class does not support streaming datasets.
+    """
+    def __init__(self, name: Optional[str] = None):
+        assert isinstance(name, basestring), "name should be a string"
+        self._name = name or str(uuid4())
+        self._jvm = None
+        self._jo = None
+
+    def _on(self, df: DataFrame, *exprs: Column) -> DataFrame:
+        """
+        Attaches this observation to the given :class:`Dataset` to observe aggregation expressions.
+
+        :param df: the `DataFrame` to be observe
+        :param exprs: aggregation expressions
+        :return: the observed `DataFrame`
+        """
+        assert exprs, "exprs should not be empty"
+        assert all(isinstance(c, Column) for c in exprs), "all exprs should be Column"
+        assert self._jo is None, "an Observation can be used with a DataFrame only once"
+
+        self._jvm = df._sc._jvm
+        self._jo = self._jvm.org.apache.spark.sql.Observation(self._name)
+        observed_df = self._jo.on(df._jdf,
+                                  exprs[0]._jc,
+                                  column._to_seq(df._sc, [c._jc for c in exprs[1:]]))
+        return DataFrame(observed_df, df.sql_ctx)
+
+    @property
+    def get(self) -> Row:
+        """
+        Get the observed metrics. This waits until the observed dataset finishes its first action.
+        Only the result of the first action is available. Subsequent actions do not modify the
+        result.
+
+        :return: the metrics :class:`Row`
+        """
+        assert self._jo is not None, 'call DataFrame.observe / Observation.on first'
+        jrow = self._jo.get()
+        return self._to_row(jrow)
+
+    def _to_row(self, jrow):
+        field_names = jrow.schema().fieldNames()
+        valuesScalaMap = jrow.getValuesMap(self._jvm.PythonUtils.toSeq(list(field_names)))
+        valuesJavaMap = self._jvm.scala.collection.JavaConversions.mapAsJavaMap(valuesScalaMap)
+        return Row(**valuesJavaMap)
```

Comment:
it also has to have a `_test` method with `if __name__ == "__main__":` in order to enable doctests. Should also add this module into https://github.com/apache/spark/blob/master/dev/sparktestsupport/modules.py#L439

Reply:
I was wondering how you are testing those code examples without defining `df` et al. Thanks for the pointers.